kaiax/reward/impl/getter.go

```diff
@@ -0,0 +1,395 @@
+
+	// Non-deferred mode
+	if !config.DeferredTxFee {
+		var proposer *big.Int
+		if config.Rules.IsMagma {
+			// In non-deferred mode, no fees to distribute here at the end of block processing.
+			// Just distribute the minting reward to the proposer and stop.
+			proposer = new(big.Int).Set(minted)
+			totalFee = big.NewInt(0)
+		} else {
+			// But Simple policy had a bug where transaction fees were distributed to the proposer here at the end of block processing
+			// despite configured to non-deferred mode. To keep the backward compatibility, the buggy behavior retains until Magma.
+			proposer = new(big.Int).Add(minted, totalFee)
+		}
+		spec.Minted = new(big.Int).Set(minted)
+		spec.TotalFee = totalFee
+		spec.BurntFee = big.NewInt(0)
+		spec.Proposer = proposer
+		spec.IncReceipient(config.Rewardbase, proposer)
+		return spec, nil
+	}
+
+	// Deferred mode
+	burntFee := big.NewInt(0)
+	if config.Rules.IsMagma {
+		burntFee = getBurnAmountMagma(totalFee)
+	}
+	proposer := new(big.Int).Add(minted, totalFee)
+	proposer.Sub(proposer, burntFee)
+
+	spec.Minted = minted
+	spec.TotalFee = totalFee
+	spec.BurntFee = burntFee
+	spec.Proposer = proposer
+	spec.IncReceipient(config.Rewardbase, proposer)
+	return spec, nil
+}
+
+// getDeferredRewardFull is for non-Simple policy.
+func getDeferredRewardFull(config *reward.RewardConfig, totalFee *big.Int, si *staking.StakingInfo) (*reward.RewardSpec, error) {
+	// Non-deferred and deferred modes share most of the logic
+	// except that in non-deferred mode the block fees are considered zero.
+	var burntFee *big.Int
+	if !config.DeferredTxFee {
+		totalFee = big.NewInt(0)
+		burntFee = big.NewInt(0)
+	} else {
+		if config.Rules.IsKore {
+			burntFee = getBurnAmountKore(config, totalFee)
+		} else if config.Rules.IsMagma {
+			burntFee = getBurnAmountMagma(totalFee)
+		} else {
+			burntFee = big.NewInt(0)
+		}
+	}
+
+	// Both non-deferred and deferred modes
+	if config.Rules.IsKore {
+		return getDeferredRewardFullKore(config, totalFee, burntFee, si)
+	} else {
+		return getDeferredRewardFullLegacy(config, totalFee, burntFee, si)
+	}
+}
+
+// getDeferredRewardFullKore is for non-Simple policy and after Kore.
+func getDeferredRewardFullKore(config *reward.RewardConfig, totalFee, burntFee *big.Int, si *staking.StakingInfo) (*reward.RewardSpec, error) {
+	var (
+		spec         = reward.NewRewardSpec()
+		minted       = new(big.Int).Set(config.MintingAmount)
+		remainingFee = new(big.Int).Sub(totalFee, burntFee)
+	)
+
+	// Distribute using RewardRatio first. Unlike Legacy, fees are not distributed here
+	// because fees are exclusively allocated to proposer. Remainder goes to KIF.
+	validators, kif, kef := config.RewardRatio.Split(minted)
+	ratioRemainder := calcRemainder(minted, validators, kif, kef)
+	kif.Add(kif, ratioRemainder)
+
+	// Further distribute using Kip82Ratio. Remainder goes to proposer.
+	proposer, stakers := config.Kip82Ratio.Split(validators)
```

Comment:
Is it always true that `proposer + stakers == validators`? I think there can be a remainder.

Reply:
Fixed in deb4cb8.